File: sdk/src/wrappers/openai.ts
/**
 * Pricing table for OpenAI models (USD per 1M tokens).
 * Source: https://openai.com/pricing (updated Apr 2025)
 */
const OPENAI_PRICING: Record<string, { input: number; output: number }> = {
  // GPT-5 (2026 Suite)
  "gpt-5.4": { input: 1.5, output: 6 },
  "gpt-5.3-codex": { input: 1.2, output: 4.8 },
  "gpt-5.2": { input: 1, output: 4 },
  "gpt-5-mini": { input: 0.1, output: 0.4 },
  // GPT-4o
  "gpt-4o": { input: 2.5, output: 10 },
  "gpt-4o-2024-11-20": { input: 2.5, output: 10 },
  "gpt-4o-2024-08-06": { input: 2.5, output: 10 },
  "gpt-4o-mini": { input: 0.15, output: 0.6 },
  "gpt-4o-mini-2024-07-18": { input: 0.15, output: 0.6 },
  // GPT-4 Turbo
  "gpt-4-turbo": { input: 10, output: 30 },
  "gpt-4-turbo-preview": { input: 10, output: 30 },
  "gpt-4": { input: 30, output: 60 },
  "gpt-4-32k": { input: 60, output: 120 },
  // GPT-3.5
  "gpt-3.5-turbo": { input: 0.5, output: 1.5 },
  "gpt-3.5-turbo-0125": { input: 0.5, output: 1.5 },
  // o-series
  "o1": { input: 15, output: 60 },
  "o1-mini": { input: 3, output: 12 },
  "o3-mini": { input: 1.1, output: 4.4 },
  // Third-party via OpenAI compatible API (DeepSeek/Kimi/GLM/Search)
  "deepseek-chat": { input: 0.14, output: 0.28 },
  "deepseek-reasoner": { input: 0.55, output: 2.19 },
  "deepseek-v4-lite": { input: 0.05, output: 0.15 },
  "moonshot-v1-8k": { input: 0.6, output: 2.5 },
  "moonshot-v1-32k": { input: 1.15, output: 8.0 },
  "glm-4": { input: 10, output: 10 },
  "glm-4.7": { input: 0.5, output: 0.5 },
  "glm-4-flash": { input: 0, output: 0 },
  "sonar": { input: 1.0, output: 1.0 },
  "sonar-pro": { input: 3.0, output: 15.0 },
  "grok-4.1": { input: 1.0, output: 4.0 },
  "llama-4-scout": { input: 0.1, output: 0.1 },
};

function calcOpenAICost(model: string, promptTokens: number, completionTokens: number): number {
  const pricing = OPENAI_PRICING[model] ?? { input: 2.5, output: 10 }; // fallback to gpt-4o price
  return (promptTokens * pricing.input + completionTokens * pricing.output) / 1_000_000;
}

/**
 * Wrap an OpenAI client instance to automatically track usage with TrackYourAPI.
 *
 * @example
 * ```ts
 * import OpenAI from 'openai'
 * import { wrapOpenAI } from '@trackyourapi/sdk'
 *
 * const openai = wrapOpenAI(new OpenAI(), {
 *   apiKey: 'qtn_xxx',
 *   endpoint: 'https://your-convex-site.convex.site'
 * })
 * ```
 */
export function wrapOpenAI<T extends OpenAILike>(
  client: T,
  options: { apiKey: string; endpoint: string; provider?: string; debug?: boolean }
): T {
  const { TrackYourAPIClient } = require("../client") as typeof import("../client");
  const tracker = new TrackYourAPIClient(options);
  const provider = options.provider ?? "openai";

  // Proxy the chat.completions.create method
  const originalCreate = client.chat.completions.create.bind(client.chat.completions);

  client.chat.completions.create = async function (...args: Parameters<typeof originalCreate>) {
    const start = Date.now();
    const result = await originalCreate(...(args as [any]));
    const latency = Date.now() - start;

    try {
      const response = result as ChatCompletionResponse;
      const model = response.model ?? (args[0] as { model?: string })?.model ?? "unknown";
      const usage = response.usage;
      if (usage) {
        const promptTokens = usage.prompt_tokens ?? 0;
        const completionTokens = usage.completion_tokens ?? 0;
        const totalTokens = usage.total_tokens ?? promptTokens + completionTokens;
        const cost = calcOpenAICost(model, promptTokens, completionTokens);

        tracker.track({
          provider,
          model,
          tokens: totalTokens,
          cost,
          latency,
        });
      }
    } catch {
      // never let tracking break the app
    }

    return result;
  };

  return client;
}

// Minimal type shims so the wrapper doesn't require openai as a hard dependency
interface OpenAILike {
  chat: {
    completions: {
      create: (...args: any[]) => Promise<any>;
    };
  };
}

interface ChatCompletionResponse {
  model?: string;
  usage?: {
    prompt_tokens?: number;
    completion_tokens?: number;
    total_tokens?: number;
  };
}

File: sdk-python/trackyourapi/wrappers/openai.py
"""
OpenAI wrapper for TrackYourAPI Python SDK.
"""

import time
from typing import Any
from ..client import TrackYourAPIClient

# Pricing: USD per 1M tokens (input, output)
_PRICING = {
    # GPT-5 (2026 Suite)
    "gpt-5.4": (1.5, 6.0),
    "gpt-5.3-codex": (1.2, 4.8),
    "gpt-5.2": (1.0, 4.0),
    "gpt-5-mini": (0.1, 0.4),
    # GPT-4o
    "gpt-4o": (2.5, 10.0),
    "gpt-4o-mini": (0.15, 0.6),
    "gpt-4-turbo": (10.0, 30.0),
    "gpt-4": (30.0, 60.0),
    "gpt-3.5-turbo": (0.5, 1.5),
    "o1": (15.0, 60.0),
    "o1-mini": (3.0, 12.0),
    "o3-mini": (1.1, 4.4),
    # Third-party via OpenAI compatible API
    "deepseek-chat": (0.14, 0.28),
    "deepseek-reasoner": (0.55, 2.19),
    "deepseek-v4-lite": (0.05, 0.15),
    "moonshot-v1-8k": (0.6, 2.5),
    "moonshot-v1-32k": (1.15, 8.0),
    "glm-4": (10.0, 10.0),
    "glm-4.7": (0.5, 0.5),
    "glm-4-flash": (0, 0),
    "sonar": (1.0, 1.0),
    "sonar-pro": (3.0, 15.0),
    "grok-4.1": (1.0, 4.0),
    "llama-4-scout": (0.1, 0.1),
}


def _calc_cost(model: str, prompt_tokens: int, completion_tokens: int) -> float:
    input_price, output_price = _PRICING.get(model, (2.5, 10.0))
    return (prompt_tokens * input_price + completion_tokens * output_price) / 1_000_000


def wrap_openai(client: Any, *, api_key: str, endpoint: str, provider: str = "openai", debug: bool = False) -> Any:
    """
    Wrap an OpenAI client to auto-track usage with TrackYourAPI.

    Example::

        from openai import OpenAI
        from trackyourapi import wrap_openai

        client = wrap_openai(OpenAI(), api_key="qtn_xxx", endpoint="https://YOUR.convex.site")

        # All calls are now tracked
        response = client.chat.completions.create(
            model="gpt-4o",
            messages=[{"role": "user", "content": "Hello!"}],
        )
    """
    tracker = TrackYourAPIClient(api_key=api_key, endpoint=endpoint, debug=debug)
    original_create = client.chat.completions.create

    def patched_create(*args: Any, **kwargs: Any) -> Any:
        start = time.monotonic()
        response = original_create(*args, **kwargs)
        latency_ms = (time.monotonic() - start) * 1000

        try:
            model = getattr(response, "model", kwargs.get("model", "unknown"))
            usage = getattr(response, "usage", None)
            if usage:
                prompt_tokens = getattr(usage, "prompt_tokens", 0) or 0
                completion_tokens = getattr(usage, "completion_tokens", 0) or 0
                total_tokens = getattr(usage, "total_tokens", prompt_tokens + completion_tokens) or 0
                cost = _calc_cost(model, prompt_tokens, completion_tokens)
                tracker.track(
                    provider=provider,
                    model=model,
                    tokens=total_tokens,
                    cost=cost,
                    latency=latency_ms,
                )
        except Exception:
            pass  # never break the app

        return response

    client.chat.completions.create = patched_create
    return client
